File: models/report.py
from sqlalchemy import Column, Integer, String, Float, DateTime, Enum as SQLEnum, ForeignKey
from sqlalchemy.orm import relationship
from datetime import datetime
import enum
from .base import Base

class ReportStatus(enum.Enum):
    PENDING = "Pending"
    VERIFIED = "Verified"
    REJECTED = "Rejected"

class DisasterType(enum.Enum):
    FLOOD = "Flood"
    FIRE = "Fire"
    EARTHQUAKE = "Earthquake"
    ACCIDENT = "Accident"
    SOS = "SOS"
    OTHER = "Other"

class Report(Base):
    __tablename__ = "reports"
    
    id = Column(Integer, primary_key=True, index=True)
    text = Column(String, nullable=False)
    latitude = Column(Float, nullable=False)
    longitude = Column(Float, nullable=False)
    disaster_type = Column(SQLEnum(DisasterType), nullable=False)
    status = Column(SQLEnum(ReportStatus), default=ReportStatus.PENDING)
    ip_address = Column(String, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    verified_at = Column(DateTime, nullable=True)
    
    # Foreign key relationships
    reported_by = Column(Integer, ForeignKey("users.id"))
    verified_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    
    # Relationships
    reporter = relationship("User", foreign_keys=[reported_by])
    verifier = relationship("User", foreign_keys=[verified_by])

File: models/Report.js
import mongoose from 'mongoose';

const reportSchema = new mongoose.Schema({
  description: {
    type: String,
    required: true,
    trim: true,
  },
  location: {
    type: {
      type: String,
      enum: ['Point'],
      default: 'Point',
    },
    coordinates: {
      type: [Number], // [longitude, latitude]
      required: true,
    },
  },
  disasterType: {
    type: String,
    enum: ['Flood', 'Fire', 'Earthquake', 'Other'],
    default: 'Other',
  },
  confidence: {
    type: Number,
    min: 0,
    max: 1,
    default: 0.5,
  },
  status: {
    type: String,
    enum: ['Pending', 'Verified', 'Rejected'],
    default: 'Pending',
  },
  isSOS: {
    type: Boolean,
    default: false,
  },
  images: [{
    type: String, // URLs to uploaded images
  }],
  reportedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  verifiedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
  },
  verificationNotes: {
    type: String,
  },
  ipAddress: {
    type: String,
  },
  userAgent: {
    type: String,
  },
}, {
  timestamps: true,
});

// Create 2dsphere index for location queries
reportSchema.index({ location: '2dsphere' });

// Create indexes for common queries
reportSchema.index({ status: 1, createdAt: -1 });
reportSchema.index({ disasterType: 1 });
reportSchema.index({ isSOS: 1 });

export default mongoose.models.Report || mongoose.model('Report', reportSchema);
